File: core/src/main/java/dev/bee/kanjianki/core/SimilarKanjiChoicePlanner.java
package dev.bee.kanjianki.core;

import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.TreeMap;
import java.util.TreeSet;

public final class SimilarKanjiChoicePlanner {
    public static final int FALLBACK_CHOICE_LIMIT = 4;

    public List<RecordsImportModels.SimilarKanjiChoiceCard> buildCandidates(
            List<RecordsImportModels.KanjiInventoryItem> inventory,
            List<RecordsImportModels.SimilarKanjiPair> pairs
    ) {
        Map<String, RecordsImportModels.KanjiInventoryItem> inventoryByKanji = inventoryByKanji(inventory);
        if (inventoryByKanji.size() < 2) {
            return Collections.emptyList();
        }

        Map<String, Set<String>> directNeighbors = directNeighbors(pairs, inventoryByKanji);
        List<RecordsImportModels.SimilarKanjiChoiceCard> out = new ArrayList<>();
        for (RecordsImportModels.KanjiInventoryItem target : inventoryByKanji.values()) {
            RecordsImportModels.SimilarKanjiChoiceCard card = choiceCard(target, directNeighbors);
            if (card != null) {
                out.add(card);
            }
        }
        return out;
    }

    private static Map<String, RecordsImportModels.KanjiInventoryItem> inventoryByKanji(List<RecordsImportModels.KanjiInventoryItem> inventory) {
        Map<String, RecordsImportModels.KanjiInventoryItem> inventoryByKanji = new TreeMap<>();
        if (inventory != null) {
            for (RecordsImportModels.KanjiInventoryItem item : inventory) {
                if (item != null && !item.kanji.isEmpty()) {
                    inventoryByKanji.put(item.kanji, item);
                }
            }
        }
        return inventoryByKanji;
    }

    private static Map<String, Set<String>> directNeighbors(
            List<RecordsImportModels.SimilarKanjiPair> pairs,
            Map<String, RecordsImportModels.KanjiInventoryItem> inventoryByKanji
    ) {
        Map<String, Set<String>> directNeighbors = new TreeMap<>();
        if (pairs != null) {
            for (RecordsImportModels.SimilarKanjiPair pair : pairs) {
                if (validPair(pair, inventoryByKanji)) {
                    directNeighbors.computeIfAbsent(pair.kanjiA, ignored -> new TreeSet<>()).add(pair.kanjiB);
                    directNeighbors.computeIfAbsent(pair.kanjiB, ignored -> new TreeSet<>()).add(pair.kanjiA);
                }
            }
        }
        return directNeighbors;
    }

    private static boolean validPair(RecordsImportModels.SimilarKanjiPair pair, Map<String, RecordsImportModels.KanjiInventoryItem> inventoryByKanji) {
        return pair != null
                && !pair.kanjiA.isEmpty()
                && !pair.kanjiB.isEmpty()
                && !pair.kanjiA.equals(pair.kanjiB)
                && inventoryByKanji.containsKey(pair.kanjiA)
                && inventoryByKanji.containsKey(pair.kanjiB);
    }

    private static RecordsImportModels.SimilarKanjiChoiceCard choiceCard(
            RecordsImportModels.KanjiInventoryItem target,
            Map<String, Set<String>> directNeighbors
    ) {
        String meaning = target.primaryMeaning.trim();
        Set<String> neighbors = directNeighbors.get(target.kanji);
        if (meaning.isEmpty() || neighbors == null) {
            return null;
        }
        TreeSet<String> choices = new TreeSet<>();
        choices.add(target.kanji);
        choices.addAll(neighbors);
        List<String> choiceList = new ArrayList<>(choices);
        return new RecordsImportModels.SimilarKanjiChoiceCard(
                target.kanji,
                meaning,
                choiceList,
                choiceSignature(choiceList)
        );
    }

    public RecordsImportModels.SimilarKanjiChoiceResult evaluateSelection(
            RecordsImportModels.SimilarKanjiChoiceCard card,
            String selectedKanji
    ) {
        if (card == null) {
            return new RecordsImportModels.SimilarKanjiChoiceResult(null, selectedKanji, false, Collections.emptyList());
        }
        String selected = selectedKanji == null ? "" : selectedKanji.trim();
        boolean correct = card.targetKanji.equals(selected);
        if (correct) {
            return new RecordsImportModels.SimilarKanjiChoiceResult(card, selected, true, Collections.emptyList());
        }
        LinkedHashSet<String> repairs = new LinkedHashSet<>();
        repairs.add(card.targetKanji);
        if (card.choices.contains(selected)) {
            repairs.add(selected);
        }
        return new RecordsImportModels.SimilarKanjiChoiceResult(card, selected, false, new ArrayList<>(repairs));
    }

    public static List<String> fallbackChoices(String targetKanji, List<RecordsImportModels.SimilarKanjiPair> pairs) {
        LinkedHashSet<String> choices = new LinkedHashSet<>();
        choices.add(targetKanji);
        if (pairs != null) {
            for (RecordsImportModels.SimilarKanjiPair pair : pairs) {
                if (pair == null) {
                    continue;
                }
                String other = pair.kanjiA.equals(targetKanji) ? pair.kanjiB : pair.kanjiA;
                choices.add(other);
                if (choices.size() >= FALLBACK_CHOICE_LIMIT) {
                    break;
                }
            }
        }
        return new ArrayList<>(choices);
    }

    public static String choiceSignature(List<String> choices) {
        TreeSet<String> sorted = new TreeSet<>();
        if (choices != null) {
            for (String choice : choices) {
                if (choice != null && !choice.trim().isEmpty()) {
                    sorted.add(choice.trim());
                }
            }
        }
        return String.join("\t", sorted);
    }
}

File: core/src/main/kotlin/dev/bee/kanjianki/core/SimilarKanjiChoicePlanner.kt
package dev.bee.kanjianki.core

import java.util.TreeMap
import java.util.TreeSet
import java.util.logging.Logger

class SimilarKanjiChoicePlanner {
    fun buildCandidates(
        inventory: List<RecordsImportModels.KanjiInventoryItem?>?,
        pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
    ): List<RecordsImportModels.SimilarKanjiChoiceCard> {
        return buildCandidates(inventory, pairs, emptyMap())
    }

    fun buildCandidates(
        inventory: List<RecordsImportModels.KanjiInventoryItem?>?,
        pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
        wrongPickCounts: Map<String, Map<String, Int>>?,
    ): List<RecordsImportModels.SimilarKanjiChoiceCard> {
        val inventoryByKanji = inventoryByKanji(inventory)
        if (inventoryByKanji.size < 2) {
            return emptyList()
        }

        val directNeighbors = directNeighbors(pairs, inventoryByKanji)
        val out = ArrayList<RecordsImportModels.SimilarKanjiChoiceCard>()
        for (target in inventoryByKanji.values) {
            val card = choiceCard(target, directNeighbors, wrongPickCounts?.get(target.kanji).orEmpty())
            if (card != null) {
                out.add(card)
            }
        }
        return out
    }

    fun evaluateSelection(
        card: RecordsImportModels.SimilarKanjiChoiceCard?,
        selectedKanji: String?,
    ): RecordsImportModels.SimilarKanjiChoiceResult {
        if (card == null) {
            return RecordsImportModels.SimilarKanjiChoiceResult(null, selectedKanji, false, emptyList())
        }
        val selected = selectedKanji?.trim() ?: ""
        val correct = card.targetKanji == selected
        if (correct) {
            return RecordsImportModels.SimilarKanjiChoiceResult(card, selected, true, emptyList())
        }
        val repairs = LinkedHashSet<String>()
        repairs.add(card.targetKanji)
        if (card.choices.contains(selected)) {
            repairs.add(selected)
        }
        return RecordsImportModels.SimilarKanjiChoiceResult(card, selected, false, ArrayList(repairs))
    }

    companion object {
        const val FALLBACK_CHOICE_LIMIT: Int = 4

        private fun inventoryByKanji(
            inventory: List<RecordsImportModels.KanjiInventoryItem?>?,
        ): Map<String, RecordsImportModels.KanjiInventoryItem> {
            val inventoryByKanji = TreeMap<String, RecordsImportModels.KanjiInventoryItem>()
            if (inventory != null) {
                for (item in inventory) {
                    if (item != null && item.kanji.isNotEmpty()) {
                        inventoryByKanji[item.kanji] = item
                    }
                }
            }
            return inventoryByKanji
        }

        private fun directNeighbors(
            pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
            inventoryByKanji: Map<String, RecordsImportModels.KanjiInventoryItem>,
        ): Map<String, Set<String>> {
            val directNeighbors = TreeMap<String, MutableSet<String>>()
            if (pairs != null) {
                for (pair in pairs) {
                    if (validPair(pair, inventoryByKanji)) {
                        directNeighbors.computeIfAbsent(pair!!.kanjiA) { TreeSet() }.add(pair.kanjiB)
                        directNeighbors.computeIfAbsent(pair.kanjiB) { TreeSet() }.add(pair.kanjiA)
                    }
                }
            }
            return directNeighbors
        }

        private fun validPair(
            pair: RecordsImportModels.SimilarKanjiPair?,
            inventoryByKanji: Map<String, RecordsImportModels.KanjiInventoryItem>,
        ): Boolean {
            return pair != null &&
                pair.kanjiA.isNotEmpty() &&
                pair.kanjiB.isNotEmpty() &&
                pair.kanjiA != pair.kanjiB &&
                inventoryByKanji.containsKey(pair.kanjiA) &&
                inventoryByKanji.containsKey(pair.kanjiB)
        }

        private fun choiceCard(
            target: RecordsImportModels.KanjiInventoryItem,
            directNeighbors: Map<String, Set<String>>,
            targetWrongPicks: Map<String, Int>,
        ): RecordsImportModels.SimilarKanjiChoiceCard? {
            val meaning = target.primaryMeaning.trim()
            val neighbors = directNeighbors[target.kanji]
            if (meaning.isEmpty() || neighbors == null) {
                return null
            }
            val choiceList = if (targetWrongPicks.isEmpty()) {
                val choices = TreeSet<String>()
                choices.add(target.kanji)
                choices.addAll(neighbors)
                ArrayList(choices)
            } else {
                confusionOrderedChoices(target.kanji, neighbors, targetWrongPicks)
            }
            return RecordsImportModels.SimilarKanjiChoiceCard(
                target.kanji,
                meaning,
                choiceList,
                choiceSignature(choiceList),
            )
        }

        /**
         * Orders the neighbor distractors by how often the learner actually
         * confused them with the target (wrong-pick count desc, then the
         * existing deterministic lexicographic order) so the most-confused
         * neighbors survive the choice-limit truncation.
         */
        private fun confusionOrderedChoices(
            targetKanji: String,
            neighbors: Set<String>,
            targetWrongPicks: Map<String, Int>,
        ): ArrayList<String> {
            val orderedNeighbors = neighbors
                .filter { it != targetKanji }
                .sortedWith(compareByDescending<String> { targetWrongPicks[it] ?: 0 }.thenBy { it })
            val choiceList = ArrayList<String>()
            choiceList.add(targetKanji)
            for (neighbor in orderedNeighbors) {
                if (choiceList.size >= FALLBACK_CHOICE_LIMIT) {
                    break
                }
                choiceList.add(neighbor)
            }
            return choiceList
        }

        @JvmStatic
        fun fallbackChoices(
            targetKanji: String,
            pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
        ): List<String> {
            val choices = LinkedHashSet<String>()
            choices.add(targetKanji)
            if (pairs != null) {
                for (pair in pairs) {
                    if (pair == null) {
                        continue
                    }
                    val other = if (pair.kanjiA == targetKanji) pair.kanjiB else pair.kanjiA
                    choices.add(other)
                    if (choices.size >= FALLBACK_CHOICE_LIMIT) {
                        break
                    }
                }
            }
            return ArrayList(choices)
        }

        @JvmField
        val LOGGER: Logger = Logger.getLogger(SimilarKanjiChoicePlanner::class.java.name)

        @JvmStatic
        fun choiceCardForSession(
            stored: RecordsImportModels.SimilarKanjiChoiceCard?,
            targetKanji: String,
            primaryMeaning: String?,
            pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
        ): RecordsImportModels.SimilarKanjiChoiceCard = choiceCardForSession(
            stored,
            targetKanji,
            primaryMeaning,
            pairs,
            null,
        )

        @JvmStatic
        fun choiceCardForSession(
            stored: RecordsImportModels.SimilarKanjiChoiceCard?,
            targetKanji: String,
            primaryMeaning: String?,
            pairs: List<RecordsImportModels.SimilarKanjiPair?>?,
            preferredConfusion: String?,
        ): RecordsImportModels.SimilarKanjiChoiceCard {
            val base = if (stored != null) {
                stored
            } else {
                // Goal 69 safety net: with the strengthened hasSimilarKanji predicate
                // (both pair endpoints in the local inventory), a card that reaches
                // the similar_kanji rung should always have a pre-built choice state,
                // so this fallback should be unreachable in practice. Warn if it
                // fires so a divergence between predicate and planner is caught.
                LOGGER.warning {
                    "SimilarKanjiChoicePlanner.choiceCardForSession fell back to on-the-fly " +
                        "choices for '$targetKanji'; expected a pre-built similar-kanji choice state."
                }
                val choices = fallbackChoices(targetKanji, pairs)
                RecordsImportModels.SimilarKanjiChoiceCard(
                    targetKanji,
                    primaryMeaning ?: "",
                    choices,
                    choiceSignature(choices),
                )
            }
            return withPreferredConfusion(base, preferredConfusion)
        }

        /** Keeps the captured wrong glyph in the bounded choice set first. */
        @JvmStatic
        fun withPreferredConfusion(
            card: RecordsImportModels.SimilarKanjiChoiceCard,
            preferredConfusion: String?,
        ): RecordsImportModels.SimilarKanjiChoiceCard {
            val confusion = preferredConfusion?.trim().orEmpty()
            if (confusion.isEmpty() || confusion == card.targetKanji) {
                return card
            }
            val limit = card.choices.size.coerceAtLeast(2).coerceAtMost(FALLBACK_CHOICE_LIMIT)
            val ordered = LinkedHashSet<String>()
            ordered.add(card.targetKanji)
            ordered.add(confusion)
            for (choice in card.choices) {
                if (ordered.size >= limit) break
                val normalized = choice.trim()
                if (normalized.isNotEmpty()) ordered.add(normalized)
            }
            val choices = ordered.toList()
            return RecordsImportModels.SimilarKanjiChoiceCard(
                card.targetKanji,
                card.primaryMeaning,
                choices,
                choiceSignature(choices),
            )
        }

        @JvmStatic
        fun choiceSignature(choices: List<String?>?): String {
            val sorted = TreeSet<String>()
            if (choices != null) {
                for (choice in choices) {
                    if (choice != null && choice.trim().isNotEmpty()) {
                        sorted.add(choice.trim())
                    }
                }
            }
            return sorted.joinToString("\t")
        }
    }
}
